### PR TITLE
@pepopowitz => [WIP] [SearchBar] Start some basic tests

### PR DESCRIPTION
**Pull request by mzikherman**
Starting off w/ some basic tests (suggestions are being rendered).

This library is (so far) proving to be sort of a pain to test. I can't figure out the right incantation of focus/change events on the input to actually render the suggestions. Very frustrating!

### CODE REVIEW

File: src/Components/Search/SearchBar.tsx [hunk 1]
```diff
@@ -179,6 +191,7 @@ export class SearchBar extends Component<Props, State> {
     return (
       <AutosuggestContainer>
         <Autosuggest
+          alwaysRenderSuggestions={typeof jest === "undefined" ? false : true}
```
Comment (mzikherman) [inline src/Components/Search/SearchBar.tsx]: This is the hack...I couldn't figure out how to render suggestions from tests (even though everything is being passed in properly). cc @damassi maybe we can pair on figuring out this lib testing wise?

File: src/Components/Search/__tests__/SearchBar.test.tsx [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  },
+}
+
+describe("SearchBar", () => {
+  const getWrapper = viewer => {
+    return mount(
+      <MockBoot breakpoint="xl">
+        <SearchBar
+          viewer={viewer as any}
+          relay={{ environment: {}, refetch: () => null } as any}
```
Comment (mzikherman) [inline src/Components/Search/__tests__/SearchBar.test.tsx]: We're using `refetch` (and that's not something we're testing currently, so this is just a stub for now).

**Comment by artsyit** [comment]
Deploy preview for *artsy-reaction* ready!

Built with commit 641f8c69fd9291d8c827d78b0070f2e569e027bb

https://deploy-preview-2010--artsy-reaction.netlify.com

File: src/Components/Search/__tests__/SearchBar.test.tsx [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+      </MockBoot>
+    )
+  }
+
+  it("displays search results", async () => {
+    const component = getWrapper(searchResults)
+    component
+      .find(Input)
+      .props()
+      .onChange({ target: { value: "percy" } } as any)
```
Comment (mzikherman) [inline src/Components/Search/__tests__/SearchBar.test.tsx]: No matter how hard i tried to type this properly `HTMLInputElement` I couldn't get it to not complain about all the _other_ missing attributes for that type. I just wanted the value anyway.

Comment (mzikherman) [reply src/Components/Search/__tests__/SearchBar.test.tsx]: This is also where doing `simulate("focus")` and `simulate("change", { target: { value: ...)` and various combinations didn't get the suggestions to render, leading to that hack above.

File: src/Components/Search/SearchBar.tsx [hunk 4]
```diff
@@ -159,7 +166,12 @@ export class SearchBar extends Component<Props, State> {
 
   renderInputComponent = inputProps => (
     <Box>
-      <Input style={{ width: "100%" }} {...inputProps} />
+      <Input
+        style={{ width: "100%" }}
+        innerRef={inputProps.ref}
+        {...inputProps}
+        ref={null}
```
Comment (damassi) [inline src/Components/Search/SearchBar.tsx]: why explicitly setting null here?

File: src/Components/Search/__tests__/SearchBar.test.tsx (hunk 3 above)
Comment (damassi) [reply src/Components/Search/__tests__/SearchBar.test.tsx]: Might be worth digging into recent Reaction PR's to look up @ds300's `simulateTyping` helper.

File: src/Components/Search/SearchBar.tsx (hunk 4 above)
Comment (mzikherman) [reply src/Components/Search/SearchBar.tsx]: Without that you get an error/warning, causing the test to fail: `Warning: Input: 'ref' is not a prop. Trying to access it will result in 'undefined' being returned`

This is from https://github.com/moroshko/react-autosuggest/issues/469

> Debugging a custom input that is a `styled-component`, not necessarily needing the ref as you do, `ref={null}` was the key

File: src/Components/Search/SearchBar.tsx (hunk 1 above)
Comment (damassi) [reply src/Components/Search/SearchBar.tsx]: Might be worth moving this directly into `AutoSuggest` -- meaning, `alwaysRenderSuggestions` prop takes normal true / false / app flow stuff -- and then from inside the component you override the result if in a test environment. That said, this _should_ just work since we're rendering inside of a browser environment (jsdom). Might be something up elsewhere.

File: src/Components/Search/__tests__/SearchBar.test.tsx (hunk 3 above)
Comment (mzikherman) [reply src/Components/Search/__tests__/SearchBar.test.tsx]: Yea, I saw that. Any combination of those methods (click/focus, setting value, etc.) doesn't cause the suggestion lib to actually render the suggestions. It's possibly something in the way that lib works? That's definitely the right way to simulate the methods. That's why I had to go with the `alwaysRenderSuggestions={typeof jest === "undefined" ? false : true}` hack. This lib is really frustrating me to test!

File: src/Components/Search/SearchBar.tsx (hunk 1 above)
Comment (mzikherman) [reply src/Components/Search/SearchBar.tsx]: if we're keeping a hack, it probably makes sense in our code/clients vs the actual lib. But I'd love to understand why this hack is needed.

File: src/Components/Search/SearchBar.tsx (hunk 4 above)
Comment (damassi) [reply src/Components/Search/SearchBar.tsx]: Hmm, yeah we might be running into some incompatibility stuff between styled-components 3 and 4. I'm going to finish this update today.

File: src/Components/Search/__tests__/SearchBar.test.tsx (hunk 3 above)
Comment (pepopowitz) [reply src/Components/Search/__tests__/SearchBar.test.tsx]: If there are async things happening in the library, we might need to "wait" for them before things get rendered.

[Here are some tests](https://github.com/artsy/reaction/blob/2a64f4028994c40e2595ada34b78e1ea9bc76d28/src/Apps/Order/Routes/__tests__/Payment.test.tsx#L105) that wait for async things to happen. 

Following the trail of that `update` function to [here](https://github.com/pepopowitz/reaction/blob/2a64f4028994c40e2595ada34b78e1ea9bc76d28/src/DevTools/RootTestPage.tsx#L17), here's what's happening: 

1. Wait for a 0-length promise to resolve, then
2. Call `update()` on the `root`, which is really just the top-level wrapper returned by `enzyme#mount`, then
3. Call `update()` on a wrapper that ...I'm not sure on this part, exactly. It might be an element that got selected during the test. 

Regardless of #3, I think it's worth trying 1&2 to see if asynchrony is the problem.

Comment (pepopowitz) [reply src/Components/Search/__tests__/SearchBar.test.tsx]: In general, I feel like using `.simulate('change')` is the right move. Using `.props().onChange` seems like a workaround for enzyme problems, and that strikes me as something that will lead to future maintenance headaches.

Comment (damassi) [reply src/Components/Search/__tests__/SearchBar.test.tsx]: Int terms of waiting, there is also the [waitFor](https://github.com/artsy/reaction/blob/master/src/DevTools/enzyme%2BrenderUntil.d.ts) util that you can tap into.

Comment (mzikherman) [reply src/Components/Search/__tests__/SearchBar.test.tsx]: No dice on the `update()` . I did switch back to `simulate` which works w/ the `alwaysRenderSuggestions` hack, so that's no issue.

Comment (mzikherman) [reply src/Components/Search/__tests__/SearchBar.test.tsx]: Alright we figured it out (sorta...). Needed a `setImmediate(` on the assertions in the test.

**Comment by mzikherman** [comment]
Ok @pepopowitz @jonallured I think this is mergeable, some meaningful tests are starting to come along!

File: src/Components/Search/__tests__/SearchBar.test.tsx [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    },
+    wrapper: children => (
+      <MockBoot breakpoint={breakpoint as any}>{children}</MockBoot>
+    ),
+  })
+}
+
+describe("SearchBar", () => {
+  it("displays search results", async done => {
+    const component = await getWrapper(searchResults)
```
Comment (jonallured) [inline src/Components/Search/__tests__/SearchBar.test.tsx]: I think it would be more clear to call this `wrapper` especially since the func is called `getWrapper`.

File: src/Components/Search/__tests__/SearchBar.test.tsx [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    setImmediate(() => {
+      expect(component.find(Input).props().placeholder).toBe(
+        "Search by artist, gallery, style, theme, tag, etc."
+      )
+      done()
+    })
+  })
+
+  it("doesn't display placeholder text in the xs breakpoint", async done => {
```
Comment (jonallured) [inline src/Components/Search/__tests__/SearchBar.test.tsx]: Nice one! 😎

File: src/Components/Search/__tests__/SearchBar.test.tsx [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    })
+  })
+
+  it("navigates the user when clicking on an item", async done => {
+    const component = await getWrapper(searchResults)
+    simulateTyping(component, "blah") // Any text of non-zero length.
+    setImmediate(() => {
+      window.location.assign = jest.fn()
+      component.find(SuggestionItem).simulate("click")
+      expect(window.location.assign).toHaveBeenCalledWith("/cat/percy-z")
```
Comment (jonallured) [inline src/Components/Search/__tests__/SearchBar.test.tsx]: 💯

**Comment by pepopowitz** [comment]
Re: `setImmediate` and `done`: 

Does `setImmediate` solve anything that `await flushPromiseQueue()` doesn't? If not, I'd switch to `flushPromiseQueue` to avoid having to use `done()`.

And if we *need* `setImmediate`, can we wrap it in an async/await? I think if we do that, we don't need to call `done()`.

**Comment by mzikherman** [comment]
oo I didn't know about that `flushPromiseQueue` helper, yea totally can use that. That's what was I was missing in these tests and wasn't able to figure out (`setTimeout` with 0/non-zero was also working, so just went with `setImmediate`).

**Comment by mzikherman** [comment]
Ok I think this is mergeable! Went with `flushPromiseQueue` helper.

**Comment by jonallured** [comment]
Nice one!! ❤️

**Comment by artsyit** [comment]
:rocket: PR was released in v12.1.33 :rocket:

**Comment by pepopowitz** [comment]
These tests look really great @mzikherman!!! 👌
